core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -1751,10 +1768,56 @@ private List<AcquiredRecords> createBatches(
                 .setLastOffset(lastAcquiredOffset)
                 .setDeliveryCount((short) 1));
 
-            result.forEach(acquiredRecords -> {
-                // Schedule acquisition lock timeout for the batch.
-                AcquisitionLockTimerTask timerTask = scheduleAcquisitionLockTimeout(memberId, acquiredRecords.firstOffset(), acquiredRecords.lastOffset());
-                // Add the new batch to the in-flight records along with the acquisition lock timeout task for the batch.
+            if (isRecordLimitMode) {
+                AcquiredRecords acquiredRecords = result.get(0);
+                if (acquiredRecords.lastOffset() - acquiredRecords.firstOffset() + 1 > maxFetchRecords) {
+                    InFlightBatch inFlightBatch = new InFlightBatch(
+                        timer,
+                        time,
+                        memberId,
+                        acquiredRecords.firstOffset(),
+                        acquiredRecords.lastOffset(),
+                        RecordState.ACQUIRED,
+                        1,
+                        null,
+                        timeoutHandler,
+                        sharePartitionMetrics);
+
+                    int delayMs = recordLockDurationMsOrDefault(groupConfigManager, groupId, defaultRecordLockDurationMs);
+                    long lastOffset = acquiredRecords.firstOffset() + maxFetchRecords - 1;
+                    acquiredRecords.setLastOffset(lastOffset);
+                    inFlightBatch.maybeInitializeOffsetStateUpdate(lastOffset, delayMs);
+                    updateFindNextFetchOffset(true);
+
+                    cachedState.put(acquiredRecords.firstOffset(), inFlightBatch);
+                    sharePartitionMetrics.recordInFlightBatchMessageCount(
+                        acquiredRecords.lastOffset() - acquiredRecords.firstOffset() + 1);
+                    return List.of(acquiredRecords);
+                } else {
+                    addBatches(memberId, result);
+                }
+            } else {
+                addBatches(memberId, result);
+            }
```

Comment:
If has a return statement anyways
```suggestion
            addBatches(memberId, result);
```

Reply:
Done, thanks for comment.